hydromt/metadata_resolvers/convention_resolver.py

```diff
@@ -0,0 +1,110 @@
+"""MetaDataResolver using HydroMT naming conventions."""
+from itertools import product
+from string import Formatter
+from typing import TYPE_CHECKING, Any, Dict, List, Optional, Set, Tuple
+
+import geopandas as gpd
+import numpy as np
+import pandas as pd
+
+from hydromt._typing import Bbox, NoDataStrategy, TimeRange
+
+from .metadata_resolver import MetaDataResolver
+
+if TYPE_CHECKING:
+    from hydromt.data_sources.data_source import DataSource
+
+
+class ConventionResolver(MetaDataResolver):
+    """MetaDataResolver using HydroMT naming conventions."""
+
+    _uri_placeholders = frozenset({"year", "month", "variable", "zoom_level"})
+
+    def _expand_uri_placeholders(
+        self,
+        uri: str,
+        time_tuple: Optional[Tuple[str, str]] = None,
+        variables: Optional[Set[str]] = None,
+    ) -> Tuple[str, List[str]]:
+        """Expand known placeholders in the URI."""
+        keys: list[str] = []
+
+        if "{" in uri:
+            uri_expanded = ""
+            for literal_text, key, fmt, _ in Formatter().parse(uri):
+                uri_expanded += literal_text
+                if key is None:
+                    continue
+                key_str = "{" + f"{key}:{fmt}" + "}" if fmt else "{" + key + "}"
```

Comment:
Totally fine to do this later, but I'm wondeirng if we can replace this whole code block with a regex.